install.sh

```diff
@@ -178,25 +235,217 @@ legacy_transformations_end() {
 
     # Remove legacy registration file (we moved it to another location)
     if [ -f "$INSTALL_DIR/autoupdate/registered" ]; then
-        rm -rf "$INSTALL_DIR/autoupdate"
+        rm -rf "$INSTALL_DIR/autoupdate" || LEGACY_TRANSFORM_FAILURES="true"
+    fi
+
+    legacy_transform_registered_repos || LEGACY_TRANSFORM_FAILURES="true"
+
+    if [ "$LEGACY_TRANSFORM_FAILURES" = "true" ]; then
+        echo "! There were legacy transform errors: check stderr"
+        return 1
+    fi
+
+    return 0
+}
+
+############################################################
+# Transform all comma-delimited global \`githooks.shared\`
+#   values into multiple git config values.
+#
+# Returns:
+#   1 when failed, 0 otherwise
+############################################################
+legacy_transform_split_global_shared_entries() {
+
+    CURRENT_LIST=$(git config --global --get-all githooks.shared)
+
+    # If we have more then one of these githooks.shared values
+    # we have surely passed this install update already.
+    if [ "$(echo "$CURRENT_LIST" | wc -l)" != "1" ]; then
+        return 0
     fi
+
+    FAILURE="false"
+
+    # If it contains a comma, split it...
+    if echo "$CURRENT_LIST" | grep -q ","; then
+
+        git config --global --unset githooks.shared
+
+        # Split it and add all back
+        IFS="$IFS_COMMA_NEWLINE"
+        for ITEM in $CURRENT_LIST; do
+            unset IFS
+
+            if [ -n "$ITEM" ]; then
+                git config --global --add githooks.shared "$ITEM" || FAILURE="true"
+            fi
+
+            IFS="$IFS_COMMA_NEWLINE"
+        done
+        unset IFS
+    fi
+
+    if [ "$FAILURE" = "true" ]; then
+        echo "! Warning: Could not migrate the global shared hook repositories setting:" >&2
+        echo "\`$CURRENT_LIST\`" >&2
+        echo " Please check \`githooks.shared\` and add all comma-speparated" >&2
+        echo " values manually by running:" >&2
+        echo "  \$ git config --global --add githooks.shared <value>" >&2
+        LEGACY_TRANSFORM_FAILURES="true"
+    fi
+
     return 0
 }
 
 ############################################################
 # Function to dispatch to all legacy transformations
-#   at the end of a install into a repository
+#   for all registered repos.
 #
 # Returns:
 #   1 when failed, 0 otherwise
 ############################################################
-legacy_transformations_repo_end() {
+legacy_transform_registered_repos() {
+
+    # Check if we need transforms for `.shared` files:
+    # Put local paths into `--local githooks.shared`.
+    # which was introduced by PR #125 right after commit ab86d2a5:
+    MOVE_PATHS="false"
+    if legacy_transform_is_ancestor \
```

Comment:
I still don't fully understand why do we need this previous commit tracking business.
With grep, you could figure out if the `.shared` file contains any entries that might be local paths, and run the transformation if it matches.

Reply:
This is true, as mentioned before, I can of course to this transformation always, do you like to do that, then I take this commit tracking bussiness aways from this transformation. I rather thought, only do this when its really needed (and its tested test119)

Reply:
I'd like to try not having commit tracking here and a sort of mini-upgrade system if we can avoid, and just run all upgrades on every install and just bail out with as less work as possible when there's nothing to do. What do you think?

Reply:
The thing is you cant show always a warning for githooks.useCoreHooksPath. So with this install method you already need commit tracking... I am not sure if it makes sense for "local path moving"
- if normal mode -> always do it for all registered repos
- githooks.useCoreHooksPath -> do commit tracking -> if yes -> show warning.

So as it is right now, its better to understand from a programming standpoint.

- do commit tracking -> if yes:
  - if githooks.useCoreHooksPath  -> show warning  
  - if normal mode -> transform all registered repos

Reply:
Yes, I agree we shouldn't always show a warning. There are easier ways to avoid that though rather than implementing a whole version tracking upgrade system in Shell. 🙂
For example, you could set a git config on the repo that has undergone this change already so you can skip it next time.

Reply:
I am not really keen on leaving the current implementation in only because I did it, but I think the following:
the commit tracking stuff is just and easy query `legacy_transform_is_ancestor` which just checks if we hopped over this very special commit which needs fixes, so its not "really" complicated. 

The argument about version tracking upgrade system in Shell is not really an argument, we have no other choice :-), 

The other thing is either we maintain the upgrade process to be as consistent as we can, thats the apporach we currently are on, or we maybe simplify even more telling the user we just break stuff and he needs to fix it. Your method about the config variable in the repo is a good one, but there are really (from a software perspective, even though the whole stuff is in shell :see_no_evil:) some good cons:

1. when you install a config variable saying we already did this transform inside the config of a repo you basically introduce a global variable somewhere else (mainly the local repository) to track if you dont need to do legacy transforms etc.
2. you basically have redundancy because "we" know on what commit SHA the release  clone is, so you need consistency among the commit SHA and the newly introduced "global", which is not so easy to keep up with, redundancy is always kind of error prone. So when you install into an new repository, do you basically also introduce this global variable "=true" such that any update does not skip it? Its not so easy =)
3. I stringly think putting state into the repo or global Git config is the wrong thing to do, we already have state, and thats the whole release clone, I think its totally enough to work with this clone. and that is what `legacy_transform_is_ancestor` achieves :-)
3. global variables are generally a bad idea... 
4. Considering the above, I think I really get into cumbersome problems because of side-effects with such an approach becaus of dumb side-effects. User tampers with the git config etc... (which is really likely)...

So my conclusion is, why not just stick to the current implemented solution, is it so bad? Its just a check and then -> transform...

These legacy transform things can be discarded, once we decide to throw them over board, meaning we continue on a new branch lets say
`master-v2` or so, and we can really delete all this "legacy :hankey:  :-)

What do you think.

Reply:
@rycus86 Ping ;-)

Reply:
I don't think git config variables are inherently bad, they are there so you can set and read variables from it.
I'd be totally happy with going through the list of registered repos on update, 1) check if they have the "legacyTrasnformationForThingHasAlreadyRun" git config set locally in that repo, 2) run the transformation if it didn't and 3) set the config variable for that repo.

> So when you install into an new repository, do you basically also introduce this global variable "=true" such that any update does skip this legacy stuff? Its not so easy =)

If it's a repo that already has an invalid `.shared` file for example, you do need to run the transformation on it (or fail properly at run time, which is what we do here already I think), but there's no harm in running the transformation on it to make sure it's in a good state. On the other hand, what if you install our hooks into a new local repo after this special commit, would we just ignore the transformation there?

> we already have state, and thats the whole release clone

I was more thinking per-repo (local) git config rather than a central configuration for githooks.

> User tampers with the git config etc... (which is really likely)...

And that's OK, it's their repo. If the local config gets unset, we'll run the transformation again and set it back.

Reply:
For the `.shared` file (local path transform) that will work, and is of course a viable solution.  For `legacy_transform_split_global_shared_entries`, setting a (global) config variable: it wont work and ist just a buggy implementation, you can really only do this once when you exactly know when and why you are doing it, and that depends on the state of the update. Thats also the reason I wanted a solution which is not different for the local path transform and "split global shared config"...
One more argument why I strongly recommend against introducing more legacy/backwards compatibilty 💩 "config variables" :
- Refactoring gets way harder because we might think about those values, we need uninstall of this values. They bumble around in the local/global config...a wuaaaa I see red =)
- But as you said for the local path transforms, this is a transformation **which can be done always basically** which wont harm the state. I can for sure do this for that part. But really without **a config variable in the local repo**.

@matthijskooijman Whats your 5c on this?

Reply:
I agree that the `legacy_transform_split_global_shared_entries` simply isn't re-entrant, so only needs to run once globally, so it needs some kind of tracking. This can indeed be either by commit upgraded from, or by setting a config variable after the migration has ran. The latter is simpler, but also means that on a clean install, you have to also set all the config values for existing migrations, since there is no point in running them.

Then, for the "move local paths" migration, some other observations:
 - This modifies files that are (potentially) tracked inside the git repository itself, which I'm a bit hesitant to do automatically at all.
 - An alternative would be to just not do the transformation at all, and rely on the error message when running hooks to tell users that something needs to happen and have them do it manually. This has the advantage of keeping the user in control of their own git-tracked files completely.
 - One can argue that if a user has to do exactly the same operation in a lot of repositories, this is annoying and it would be worthwile to automate this. However, I wonder how common this situation really is. One of the reasons we forbid local path in repo-tracked `.shared` files is that it doesn't really make sense to publish local paths in there, so I would not be surprised if this situation does not really occur in practice much if at all (but maybe you have different experience). The only exception is maybe for repositories that do not have a `.shared` in the git repo itself, just one in the local clone created by the local user (since until now, this was the only way to have locally-managed per-repo hooks, right?). But again, I wonder how common these really are.
 - If we do want to do automatic migration and track whether it already happened, we really have to track this per-repo. You could do it globally and even if you have a complete list of all local repos at migration time, a user might create a new repo clone *after* migration, that also still contains a `.shared` file that needs to be migrated.
 - If we do want do automatic migration to make things easier for the user, we could also make it semi-automatic: Allow the user to run e.g. `githooks migrate local-paths` to do the heavy lifting of the migration, but never run it completely automatic. This means that only the error message is enough, and then the user can decide to migrate manually or call this command, no tracking needed at all.

Weighing the above, I would suggest to *not* do any automatic migration for the local paths, and maybe add the semi-automatic migration command (I'm not sure if it's really needed, but since we already have the code and it wouldn't hurt, why not).

Reply:
Very good points, it seems that the current implementation is a bit heavy, and all other solution are heavy weight and cumbersome also -> I like the way with `git hooks migrate local-paths` the best. We just warn the user on install with the note of running this command, and then move the implementation to the cli.sh, which can do this stuff... there
Automatic things are nice, but always have the risk of failing badly :-)
@rycus86 ?

Reply:
Ah @matthijskooijman: The semi-automatic workflow is nice, but so far bit a lot of a hassle, lot more duplicate functions and stuff which dangles into `cli.sh`.
I rather leave it automatic, we can argue, that githooks also changes global variables, which might then have side-effects, shared hooks no more beeing run etc. Which is far more heavy than changing a `.shared` file which is mostly tracked.
I am leaving the changes with the latest commit.

note for me: Detecting when to show the warning is orthogonal to how we solve it (automatic or semi automatic).

Reply:
> Ah @matthijskooijman: The semi-automatic workflow is nice, but so far bit a lot of a hassle, lot more duplicate functions and stuff which dangles into cli.sh.

That might just mean that lack of code sharing between `cli.sh` and `base-template.sh` is an issue that needs to be solved, if it hinders further improvements.

> that githooks also changes global variables, which might then have side-effects, shared hooks no more beeing run etc. 

But those global variables are githook-managed usually, which makes it less heavy in my intuition. But I see your point.